*: replace syscall with unix

vrothberg opened the pull request:
Some parts of the code required some more work than pure substitutions
as the unix interfaces expect a unix.Stat_t, so we couldn't just use
os.FileInfo's Sys() interface to get actual metadata.

Note: this PR bases on and extends https://github.com/openSUSE/umoci/pull/128
Closes #128

cyphar commented:
[You need to `go fmt` your code.](https://travis-ci.org/openSUSE/umoci/jobs/252830101#L1021)

vrothberg commented on pkg/unpriv/unpriv.go:
```diff
@@ -315,6 +320,7 @@ func Remove(path string) error {
 // currently have enough access bits.
 func RemoveAll(path string) error {
 	return errors.Wrap(Wrap(path, func(path string) error {
+		// XXX(VR): os.RemoveAll(path) works, but I am not sure if that's right
```
So, this if we use `os.RemoveAll(path)` here, the CI succeeds. Would this be an alternative here, @cyphar?

vrothberg commented on pkg/unpriv/unpriv.go:
```diff
@@ -315,6 +320,7 @@ func Remove(path string) error {
 	return errors.Wrap(Wrap(path, func(path string) error {
+		// XXX(VR): os.RemoveAll(path) works, but I am not sure if that's right
 		// If remove works, we're done.
 		err := os.Remove(path)
```
So, if we use os.RemoveAll(path) here, the CI succeeds. Would this be an alternative here, @cyphar?

cyphar commented on pkg/unpriv/unpriv.go:
```diff
@@ -58,20 +58,16 @@ func splitpath(path string) []string {
 // isNotExist tells you if err is an error that implies that either the path
 // accessed does not exist (or path components don't exist).
 func isNotExist(err error) bool {
```
The failure happens because of this change. In particular, `err == unix.ENOTDIR` will always be false. This is what I was referring to:

```go
// isNotExist tells you if err is an error that implies that either the path
// accessed does not exist (or path components don't exist).
func isNotExist(err error) bool {
	if os.IsNotExist(errors.Cause(err)) {
		return true
	}
	// Check that it's not actually an ENOTDIR.
	perr, ok := errors.Cause(err).(*os.PathError)
	if !ok {
		return false
	}
	errno := perr.Err
	return errno == syscall.ENOTDIR || errno == syscall.ENOENT
}
```

(And then just swap `syscall` for `unix`). However, I was thinking about this some more and maybe we should do this (we keep `os.IsNotExist` to avoid possible cases where the stdlib changes in some way).

```go
// isNotExist tells you if err is an error that implies that either the path
// accessed does not exist (or path components don't exist).
func isNotExist(err error) bool {
	if os.IsNotExist(errors.Cause(err)) {
		return true
	}

	// Check that it's not actually an ENOTDIR in disguise. This is technically
	// a dup of doing os.IsNotExist, but it doesn't hurt to be thorough.
	var errno error
	switch err := errors.Cause(err).(type) {
	case *os.PathError:
		errno = err.Err
	case *os.LinkError:
		errno = err.Err
	case *os.SyscallError:
		errno = err.Err
	}
	return errno == syscall.ENOTDIR || errno == syscall.ENOENT
}
```

Also, make sure you don't remove `errors.Cause`s. It's a no-op for non-wrapped error messages, but for wrapper error messages it makes sure that things like `os.IsNotExist` actually do the right thing.

vrothberg commented:
After some CI hiccups, it has finally passed \o/

cyphar commented:
Yeah, the CI hiccups were due to openSUSE repos being broken yesterday. I'll go through a review later today.

vrothberg commented:
I rebased and also fixed a copy-paste error in `utime_linux_test.go` that caused random fails since the the nano seconds of atime has been used for the mtime timestamp.

cyphar commented:
LGTM.

cyphar commented:
LGTM
